clientconn.go

```diff
@@ -963,6 +961,21 @@ func (ac *addrConn) resetTransport(resolveNow bool) {
 			ac.mu.Unlock()
 		}
 
+		ac.mu.Lock()
+		if ac.state == connectivity.Shutdown {
+			ac.mu.Unlock()
+			return
+		}
+
+		// If we're at the beginning of the address list
+		// And we didn't just see a successful handshake
+		// And it's not the very first addr to try (covered by the above if)
+		if ac.addrIdx == len(ac.addrs)-1 && !ac.successfulHandshake {
+			ac.updateConnectivityState(connectivity.TransientFailure)
```

Comment:
Is this the only time we'll set TransientFailure?  We want to make sure we set it whenever the current state is Ready and a connection error occurs, which I believe this will not do.

Reply:
Ah - I didn't know that. A few questions:

- Is there a definitive flow diagram somewhere I can consult? I've looked around in grpc/grpc-go/Documentation and grpc/grpc/doc and didn't see anything at first glance.

- The flow diagram would probably answer this, but is this the correct set of cases where we should transition into TRANSIENT FAILURE:

  ```
  - Any time a READY connection experiences a retryable error (onClose and onGoAway).
  - Any time we try and fail to dial the full list of addresses.
   ```

^ if the above is correct, I think I can do:

```
if ac.state == connectivity.Ready || (ac.addrIdx == len(ac.addrs)-1 && ac.state == connectivity.Connecting) {
  ac.updateConnectivityState(connectivity.TransientFailure)
  ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
}
```

This would cover:

- State is ready (if we got to the beginning of the retry loop with READY, that means onGoAway or onClose occurred)
- State is connecting and we reached the end of the address list

Reply:
> Is there a definitive flow diagram somewhere I can consult? I've looked around in grpc/grpc-go/Documentation and grpc/grpc/doc and didn't see anything at first glance.

There is a table of valid transitions for the channel (ClientConn) (but not a sub-channel (addrConn)) in grpc/doc, but not a full flow control diagram.  In fact, this particular situation doesn't happen in the other languages.  I have a meeting set up for later this week to discuss it with the other language leads.

What you're suggesting behavior-wise seems to make sense.

That said, if what you're doing now is the same as what we were doing before this change, then I think we should leave it so this can be purely a bug fix.

Reply:
Done - it seems to do so in 8ce7e6babb5ad539d8448d25aaba0708af2626ed. Also, put the logic I suggested in with a minor addition of making sure we're also checking for first attempt.